File: sdk/node/src/production-validation.ts
/**
 * CAP-PRODUCTION authoritative identity validation (admission-layer subset).
 *
 * Mirrors sdk/go/production_validation.go's ValidateIdentityBinding. Broader
 * cross-mirror identity normalization is task-a13f83fa step-9 scope; this
 * module covers only what the raw-packet admission layer (step-7) needs.
 */

export class IdentityMismatchError extends Error {}

interface IdentityBindingView {
  tenantId?: string;
  principalId?: string;
  actorId?: string;
  delegationId?: string;
}

interface JobRequestView {
  tenantId?: string;
  principalId?: string;
  meta?: { tenantId?: string; actorId?: string };
  env?: Record<string, string>;
  identity?: IdentityBindingView;
}

function sameIdentity(a?: IdentityBindingView, b?: IdentityBindingView): boolean {
  if (!a || !b) return false;
  return (
    a.tenantId === b.tenantId &&
    a.principalId === b.principalId &&
    a.actorId === b.actorId &&
    a.delegationId === b.delegationId
  );
}

export function validateIdentityBinding(request: JobRequestView, authoritative?: IdentityBindingView): void {
  if (!request || !authoritative || !authoritative.tenantId) {
    throw new IdentityMismatchError("missing request or authoritative identity");
  }
  const mirrors: Record<string, string | undefined> = {
    tenant: request.tenantId,
    principal: request.principalId,
    "meta.tenant": request.meta?.tenantId,
    "meta.actor": request.meta?.actorId,
    "env.tenant": request.env?.["tenant_id"],
    "env.principal": request.env?.["principal_id"],
  };
  for (const [name, actual] of Object.entries(mirrors)) {
    if (!actual) continue;
    let expected: string | undefined = authoritative.tenantId;
    if (name.includes("actor")) expected = authoritative.actorId;
    else if (name.includes("principal")) expected = authoritative.principalId;
    if (actual !== expected) {
      throw new IdentityMismatchError(`${name} mismatch`);
    }
  }
  if (request.identity && !sameIdentity(request.identity, authoritative)) {
    throw new IdentityMismatchError("request.identity mismatch");
  }
}

File: sdk/python/cap/production_validation.py
"""CAP-PRODUCTION authoritative identity validation (admission-layer subset).

Mirrors sdk/go/production_validation.go's ValidateIdentityBinding. Broader
cross-mirror identity normalization (top-level/meta/env/label agreement
across every ingress) is task-a13f83fa step-9 scope; this module covers only
what the raw-packet admission layer (step-7) needs before dispatch.
"""

from __future__ import annotations

from .pb.cordum.agent.v1.job_pb2 import IdentityBinding, JobRequest


class IdentityMismatchError(ValueError):
    """A JobRequest mirror disagrees with the authoritative IdentityBinding."""


def validate_identity_binding(request: JobRequest, authoritative: IdentityBinding) -> None:
    if request is None or authoritative is None or not authoritative.tenant_id:
        raise IdentityMismatchError("missing request or authoritative identity")
    mirrors = {
        "tenant": request.tenant_id,
        "principal": request.principal_id,
        "meta.tenant": request.meta.tenant_id if request.HasField("meta") else "",
        "meta.actor": request.meta.actor_id if request.HasField("meta") else "",
        "env.tenant": request.env.get("tenant_id", ""),
        "env.principal": request.env.get("principal_id", ""),
    }
    for name, actual in mirrors.items():
        if not actual:
            continue
        if "actor" in name:
            expected = authoritative.actor_id
        elif "principal" in name:
            expected = authoritative.principal_id
        else:
            expected = authoritative.tenant_id
        if actual != expected:
            raise IdentityMismatchError(f"{name} mismatch")
    if request.HasField("identity") and not _same_identity(request.identity, authoritative):
        raise IdentityMismatchError("request.identity mismatch")


def _same_identity(a: IdentityBinding, b: IdentityBinding) -> bool:
    return (
        a.tenant_id == b.tenant_id
        and a.principal_id == b.principal_id
        and a.actor_id == b.actor_id
        and a.delegation_id == b.delegation_id
    )
